### ISSUE TITLE
RFC: RxViewModel (or RxEngine)

### ISSUE THREAD

**Issue by yjaaidi** [rx-angular/rx-angular#1370]
# TL;DR

Move from this:

```ts
interface State {
 count?: number;
}

@Component({
  providers: [RxState],
  template: `
    {{ count$ | push }}
    <button (click)="increment()">+</button>
  `
})
class Counter {
  count$ = this._state.select('count');

  constructor(private _state: RxState<State>) {
   this._state.set({count: 0});
  }

  increment() {
    this._state.set(({count}) => ({count: count + 1});
    // or this._state.set({count: this._state.get('count') + 1});
  } 

}
```

to this:

```ts
const {provideRxEngine, injectRxEngine} = describeViewModel({
    count: 0
  }, {
    strategy: viewportStrategy
  }
);

@Component({
  providers: [provideRxEngine()],
  template: `
    {{ vm.count }}
    <button (click)="vm.count = vm.count + 1">+</button>
  `
})
class Counter {
  engine = injectRxEngine();
  vm = engine.vm;
}
```

# Motivation

The combination of `@rx-angular/state` and `@rx-angular/template` provide great fine-grained APIs for implementing efficient reactive components.
The goal of this RFC is to provide a coarse-grained API in order to:
1. **reduce the boilerplate,**
2. **glue on `@rx-angular/state` & `@rx-angular/template`**,
3. **provide a progressive approach where developers can dive slowly into the fine-grained APIs**

For instance, in order to increment a simple counter, current code would look something like this:
```ts
interface State {
 count?: number;
}

@Component({
  providers: [RxState],
  template: `
    {{ count$ | push }}
    <button (click)="increment()">+</button>
  `
})
class Counter {
  count$ = this._state.select('count');

  constructor(private _state: RxState<State>) {
   this._state.set({count: 0});
  }

  increment() {
    this._state.set(({count}) => ({count: count + 1});
    // or this._state.set({count: this._state.get('count') + 1});
  } 

}
```

## Improve the provide/inject

First, we can already reduce some of the boilerplate using the Private Store pattern: https://marmicode.io/blog/angular-private-services

```ts
const {provideState, injectState} = describeState({count: 0}); // The infered typed will be `{count: number}` instead of `{count?: number}.

@Component({
  providers: [provideState()],
  template: `
    {{ count$ | push }}
    <button (click)="increment()">+</button>
  `
})
class Counter {
  count$: Observable<number>;
  private _state = injectState();

  constructor() {
   this.count$ = this._state.select('count');
  }

  increment() {
    this._state.set(({count}) => ({count: count + 1});
    // or this._state.set({count: this._state.get('count') + 1});
  } 

}
```

## Trigger Change Detection

We can reduce the boilerplate by triggering the change detection when the state changes.
It's kind of a heuristic because we assume that CD is probably required when the state changes while it might be unnecessary if the value is not used of course but this can extremely simplify the code.

Now the problem is that `RxState` shouldn't get coupled to the change detection.
We probably need a thin wrapper or something. I've been thinking of `RxViewModel`. Not a big fan of the name but couldn't find any better idea yet, so I'm counting on you 😉

```ts
const {provideViewModel, injectViewModel} = describeViewModel({
    count: 0
  }, {
    strategy: viewPortStrategy // or any strategy for controlling coalescing and scheduling.
  });

@Component({
  providers: [provideViewModel()],
  template: `
    {{ state.get('count') }}
    <button (click)="increment()">+</button>
  `
})
class Counter {
  vm = injectViewModel();

  increment() {
    this._state.set(({count}) => ({count: count + 1});
    // or this._state.set({count: this._state.get('count') + 1});
  } 

}
```

`RxViewModel` will trigger the change detection anytime the state changes and based on the coalescing / scheduling strategies that can even be provided to the `describeViewModel` function.

## Setters and getters

Finally, this could be simplified even more if we provide setters/getters using a Proxy and it could look something like this:

```ts
@Component({
  providers: [provideViewModel()],
  template: `
    {{ vm.count }}
    <button (click)="vm.count = vm.count + 1">+</button>
  `
})
class Counter {
  vm = injectViewModel();
}
```
the main challenge here is collisions between the state field and `RxViewModel` fields & methods like `connect` or `hold`. 
We could accept the risk and prohibit the fields in the state but each time we add a field/method to `RxViewModel` it might break for apps that have been using that field in their state.
We could also simply name this differently and have a `vm` property:
```ts
class Counter {
  engine = injectRxEngine();
  vm = engine.vm;
}
```

**Comment by hoebbelsB** [comment]
hi @yjaaidi i'm a huge huge fan of your idea!

While reading your proposal some questions raised in my mind. 

### Improve the provide/inject

The private store pattern is only available with angular 14, right? So we would need to migrate, thus losing backward compatibility for older angular version. At least that's what we've experienced earlier and is the reason for why we are still on v12.

It shouldn't be a dealbreaker though. Just wanted to be sure about it :)

### Trigger Change Detection

YES :rocket: !!!! That's a really great one and I love the idea. 
Giving developers more control over when change-detection is executed is super amazing. I think it will even embrace devs to build smaller-sized components.

There are some unmentioned challenges though I think. This pattern is great for 'mutable' components, like your example clearly showcases.

It will get a bit more tricky when thinking about `@Input` bindings. Consider the following example:

```ts
// counter.component.ts

@Component({
  providers: [provideViewModel()],
  template: `
    {{ vm.count }}
  `
})
class Counter {
  vm = injectViewModel();
  @Input()
  set count(c: number) {
   vm.count = c;
  }
}

```

```ts
// container-component.ts

@Component({
  providers: [provideViewModel()],
  template: `
    <counter [count]="vm.count"></counter>
  `
})
class Container {
  vm = injectViewModel();
  
}

```

There are 2 components using `RxStategyCredentials` to schedule local change-detection. 
The `Container` component schedules a cd run in order to update `Counter` component.
When `Counter` component receives the input value, it will again schedule a task to render the changed value.

Imo the update passed as `Input` should be bypassed somehow. Especially leaf components which are deeply nested in a hierarchy would receive changes very late and could lead to annoying user experiences.

Do you have any idea how one could potentially tackle this problem? Btw. the same issue exists when there are many levels of nested `rxLet` directives.

### Setters and getters

I like `RxEngine` :+1: .

Could you just enlighten me what you mean with that? I didn't get the problem you describe ^^
> the main challenge here is collisions between the state field and RxViewModel fields & methods like connect or hold.
We could accept the risk and prohibit the fields in the state but each time we add a field/method to RxViewModel it might break for apps that have been using that field in their state.

**Comment by yjaaidi** [comment]
> The private store pattern is only available with angular 14, right? So we would need to migrate, thus losing backward compatibility for older angular version. At least that's what we've experienced earlier and is the reason for why we are still on v12.
> It shouldn't be a dealbreaker though. Just wanted to be sure about it :)

For previous versions, I think that we can call `ɵɵdirectiveInject`. After all, we don't care if it disappears later as since Angular 14 we have `inject()` available in the constructor.

> It will get a bit more tricky when thinking about @Input bindings.

Oh! That's a good one. This will depend a lot on how the strategy works.
Maybe the default strategies would simply trigger CD in sync if changes happen during the parent's CD or something like that.
Also, we could discourage prop drilling to avoid this "known issue".

> Could you just enlighten me what you mean with that? I didn't get the problem you describe ^^

What I meant is that I thought it would be better to put `vm` inside the `Engine` because if we merge both objects in `ViewModel` then if we add a `vm.strategy` property, users will not be able to have a property named `strategy` in their state. I just think that it's better to separate the `ViewModel` class from the thing that handles the strategy... not sure if I am clear now 😅. Let me know and I'll come up with an example.

**Comment by mhamri** [comment]
I liked everything other than the change detection part. in a matter of fact I already planned to use what you've suggested in your blog but if you ask me to choose between magic and verbosity I always go for verbosity. 

it should be clear where in the pipeline we are going to receive an update and where we are going to trigger a change detection. usually we make sure it's the last part in the pipeline, but with this it becomes totally blurry where in the pipeline we trigger change detection. 

also I was looking at `@ngneat/elf` and I liked how it let you create multiple store and I think this proposal is solving that issue perfectly. but change detection part is a complete no no for me

**Comment by yjaaidi** [comment]
Hi @mhamri, that's the reason why I separated Engine/ViewModel from State.
They are the glue taking care of triggering change detection when necessary.
Nothing more magic or implicit than native CD or OnPush.
It's probably just a matter of naming.

What do you think?

**Comment by mhamri** [comment]
@yjaaidi don't get me wrong your idea is awesome 

so you basically you want to have two functions `describeState` and `describeViewModel`
```
const {provideState, injectState} = describeState({count: 0});
const {provideViewModel, injectViewModel} = describeViewModel({
    count: 0
  }, {
    strategy: viewPortStrategy // or any strategy for controlling coalescing and scheduling.
  });
```

but I'm against the `describeViewModel` (`describeRxEngine`), but `describeState` is what I love about it. I just give you an example why `describeViewModel` is a bad idea. 

currently for my side hustle I'm writing an Ionic app and I'm using rx-angular heavily. in mobile the performance is very important and you get a sense of bad practice where you probably won't get that in web.  rx-angular is like a blessing for it, that's why I even pass observables as the `@Input` to components. there is no where in the app that I pass raw values. to be frank the feeling is like 60FPS and it's all thanks to rx-angular. I use rx-angular even as state management for the whole app. though it has some limitation so that's why I was looking at elf. because I use rx-angular as app state manger the state really got messy, that's why `describeState` is perfect to solve that issue. as it's a mobile app for performance reasons I have to make sure the subscribe only happens on the template. calling change detection on the component side is a performance killer. 

firstly async or push pipe is more clear about what type of data I'm dealing with through the app and I can change it to a stream if I need to. secondly async pipe let me to have partial rendering instead of the whole component rendering. in one template I have like 3 streams of data and each has it's own async pipe. this helps me very much to prevent re-render the whole component. but if it was a single object instead of 3 streams, I had to render the whole component on any single value changes. 

I managed the performance with two approaches: as many async pipe as possible at template side so each pipe cause a partial render. and not having anything at component that cause a change detection. 

but with `describeViewModel` we are creating a convenient way to cause change detection which is hidden from the developer. it has side effect and side effects could be big. so that's why for something like change detection I won't suggest to hide it inside a helper function. when we set a component to `OnPush` then we clearly know what could cause a component render. but with `describeViewModel` that becomes unclear. any value changes on state could cause a render, maybe I need to keep one value until I receive the second one and then I want render the component. to achieve this what should I do to prevent an unwanted render? if I can't keep data in the state i need to have local variables to keep it and if my component isn't `OnPush` still it cause a change detection. so basically `describeViewModel` didn't prevent me for not creating my component as `OnPush`. it just made things more complex.

**Comment by hoebbelsB** [comment]
@yjaaidi let's bring this back on the plate as I think it's an amazing proposal!

> What I meant is that I thought it would be better to put vm inside the Engine because if we merge both objects in ViewModel then if we add a vm.strategy property, users will not be able to have a property named strategy in their state. I just think that it's better to separate the ViewModel class from the thing that handles the strategy... not sure if I am clear now sweat_smile. Let me know and I'll come up with an example.

Okay, i totally see the point! I actually like the idea of having a `vm` proxy exposed by the engine.

I've actually played around with that idea and the results are pretty promising. I think this feature can be achieved with very little code and would have a high impact and make it easier for people to apply fine-grained change detection to their applications.

The `@Input` issue i've mentioned before shouldn't be an issue, as the `RxEngine` should `detach` the component anyway. There are just some things that are still a bit of a riddle to me:
* The timing when to run change-detection initially (OnInit ?)
* What about `@ViewChildren` and `@ContentChildren`?
  * do we run change-detection on NgAfterViewInit and NgAfterContentInit to update the queries?
* what about HostBindings? They don't work well with detached components, as they are not re-evaluated when calling `.detectChanges` and `detach` makes the component basically a cd-shield for changes from "above"
  * we can introduce custom logic for it (e.g. `describeViewModel(initialState, { host: '' })`
  * we can use [`RxStyle`](https://github.com/rx-angular/rx-angular/pull/1465) or something similar 
*  finally, as u already mentioned: how to actually glue state and engine/viewModel

okay, as I said before, i've played around with it: so [here is a branch](https://github.com/rx-angular/rx-angular/tree/poc-rx-engine) with a very simple PoC implementation [of just the ViewModel/Engine](https://github.com/rx-angular/rx-angular/commit/e694012c7441773e2cf885f18f4891ab263f5fae#diff-867865cee2476c658f5097517d38d9843c95d61a6228d53b2451441eb6d3f29e)

Let me know what you think and if we should continue together on that

**Comment by yjaaidi** [comment]
Hey Julian, so sorry for the late reply.

> The @Input issue i've mentioned before shouldn't be an issue, as the RxEngine should detach the component anyway. Are you sure? In cases of prop drilling, the change will cascade component per component at each "cycle" depending on the component's strategy.

**TL;DR: Accept it as known limitation and discourage prop drilling**

If A -> B -> C are chained in this order and they are all using `requestAnimationFrame` strategy, this won't coalesce and will need 3 animation frames to propagate to C. Right?
That said, if they the coalescing strategy is global, we could execute everything instantly if changes happen why applying the coalesced changes... somehow like it's already done by Angular.
Though, this could be added later... or known as a limitation discouraging prop drilling. The latter is my favorite.

> The timing when to run change-detection initially (OnInit ?)

Yes, I think that this would be the most unsurprising behavior.

Of course, if inputs are set before `OnInit`, they might trigger change detection before `OnInit` depending on the strategy but it's not surprising either as `OnChanges` might be triggered before `OnInit`.

> What about `@ViewChildren` and `@ContentChildren`?

Good point! I didn't think about it.
I'd say CD on `NgAfterViewInit` (to initialize the `@ViewChildren` which will be updated by future CDs so no need for CD on `NgAfterViewChecked`) and `NgAfterContentChecked` (we will want to update `@ContentChildren` each time the content is checked and this should be trigger when CD is ran on the parent holding the content, right?)

> what about HostBindings? They don't work well with detached components, as they are not re-evaluated when calling 
.detectChanges and detach makes the component basically a cd-shield for changes from "above"

Wow! Are you sure about `detectChanges` not updating `HostBindings`?

👉 This: _we can introduce custom logic for it (e.g. describeViewModel(initialState, { host: '' })_
Less boilerplate and easy to discover.

> finally, as u already mentioned: how to actually glue state and engine/viewModel

I'd say something like this but I might be missing something.
`vm.hold(vm.select(), () => engine.markForCheck())`

> okay, as I said before, i've played around with it: so [here is a branch](https://github.com/rx-angular/rx-angular/tree/poc-rx-engine) with a very simple PoC implementation [of just the ViewModel/Engine](https://github.com/rx-angular/rx-angular/commit/e694012c7441773e2cf885f18f4891ab263f5fae#diff-867865cee2476c658f5097517d38d9843c95d61a6228d53b2451441eb6d3f29e)

Amazing! I'll check it out right now! Thank you so much for putting this together.



@hoebbelsB

**Comment by hoebbelsB** [comment]
> If A -> B -> C are chained in this order and they are all using requestAnimationFrame strategy, this won't coalesce and will need 3 animation frames to propagate to C. Right?
That said, if they the coalescing strategy is global, we could execute everything instantly if changes happen why applying the coalesced changes... somehow like it's already done by Angular.
Though, this could be added later... or known as a limitation discouraging prop drilling. The latter is my favorite.

ha, that's a good one. I think this is a very similar situation as having nested `rxLet` directives, e.g.

```html
<div *rxLet="state$; let state">
   <div *rxLet="nestedState$; let nestedState"></div>
</div>
```

The good thing is, even this setup doesn't necessarily mean that you always end up in a waterfall like rendering model. Let me try to put the behavior into some words.

The following explanation is only true when rendering with `RxConcurrentStrategies` (or using our internal `scheduler`). In general, the first `rxLet` directive will spawn a new task for creating the nested `rxLet` directive and this setup already sounds like an issue. However, the task for creating the nested `rxLet` directive is put into the schedulers queue while exhausting it, the process is not finished and is able to pick up more tasks in between. 
So as the first `rxLet` renders it will put the nested directive as new task into the queue. Now there are multiple parameters affecting the outcome of this scenario:
* configured FPS
* configured rendering strategy for the nested let directive
* amount of work to create the outer rxLet
* amount of time already spent exhausting the current scheduler queue
* amount of other tasks with higher priority put into the queue :shrug: 

If the inner `rxLet` is configured with `strategy: 'immediate'` you can be sure that it will be picked up in the same task -> you will never face waterfall rendering.

If the outer `rxLet` has literally no workload, the inner rxLet probably will be rendered regardless of the used strategy.

> Good point! I didn't think about it.
I'd say CD on NgAfterViewInit (to initialize the @ViewChildren which will be updated by future CDs so no need for CD on NgAfterViewChecked) and NgAfterContentChecked (we will want to update @ContentChildren each time the content is checked and this should be trigger when CD is ran on the parent holding the content, right?)

CD after `NgAfterViewInit` sounds good (ofc there will be edge cases I think. we should have some unit tests for this). Don't you think running change detection on `NgAfterContentChecked` is a bit brutal? I think contentchildren will be the hardest part to fully support with all edge cases...

> Wow! Are you sure about detectChanges not updating HostBindings?

YES! You can re-check in my PoC repo. The usage of HostBindings is implemented, but commented out. You should be able to double-check my theory there :D 

> I'd say something like this but I might be missing something.
vm.hold(vm.select(), () => engine.markForCheck())

This looks like something that is happening internally inside of the engine, yes. But I was thinking about how users populate their state into the vm.

e.g.

```ts
vm.connect(state.select());

vm.connect('prop', someObservable$);
```
